src/sunmatrix/dense/sunmatrix_dense.c

```diff
@@ -288,20 +297,26 @@ int SUNMatScaleAdd_Dense(sunrealtype c, SUNMatrix A, SUNMatrix B)
     }
   }
 
-  return SUNMAT_SUCCESS;
+  return SUN_SUCCESS;
 }
 
-int SUNMatMatvec_Dense(SUNMatrix A, N_Vector x, N_Vector y)
+SUNErrCode SUNMatMatvec_Dense(SUNMatrix A, N_Vector x, N_Vector y)
 {
+  SUNFunctionBegin(A->sunctx);
   sunindextype i, j;
   sunrealtype *col_j, *xd, *yd;
 
-  if (!compatibleMatrixAndVectors(A, x, y)) { return SUNMAT_ILL_INPUT; }
+  SUNAssert(SUNMatGetID(A) == SUNMATRIX_DENSE, SUN_ERR_ARG_WRONGTYPE);
+  SUNCheck(compatibleMatrixAndVectors(A, x, y), SUN_ERR_ARG_DIMSMISMATCH);
 
   /* access vector data (return if NULL data pointers) */
   xd = N_VGetArrayPointer(x);
+  SUNCheckLastErr();
   yd = N_VGetArrayPointer(y);
-  if ((xd == NULL) || (yd == NULL) || (xd == yd)) { return SUNMAT_MEM_FAIL; }
+  SUNCheckLastErr();
+
+  SUNAssert(xd, SUN_ERR_MEM_FAIL);
+  SUNAssert(yd, SUN_ERR_MEM_FAIL);
```

Comment:
We previously ensured that `xd` and `yd` did not point to the same memory.  That check has been removed.
```suggestion
  SUNAssert(xd, SUN_ERR_MEM_FAIL);
  SUNAssert(yd, SUN_ERR_MEM_FAIL);
  SUNAssert(xd != yd, SUN_ERR_MEM_FAIL);
```